### DISCUSSION TITLE
Changing entity order

### DISCUSSION THREAD

**Discussion by hfn92** [skypjack/entt#1058] [Q&A]
Hi, I have a case where I want to reorder some components after using `sort` while otherwise retaining the sorted order. Is this possible with the current api?

I was hoping to be able to something like this

```cpp
  esc.sort<ToSort>([](auto& l, auto& r) { return l < r; });

  for (auto&& [e, order] : ecs.view<ForceOrderBefore>().each())
  {
    auto it = ecs.storage<ToSort>().find(order.entity);
    ecs.storage<ToSort>().move_to(it, e); // reinsert?
  }

  for (auto&& [e, order] : ecs.view<ForceOrderAfter>().each())
  {
    auto it = ecs.storage<ToSort>().find(order.entity);
    ecs.storage<ToSort>().move_to(it + 1, e);
  }
```

**Comment by skypjack** [comment]
Elements are stored in a paged vector. `move_to` isn't **that** easy to implement actually. 🙂 
However, why not a single `ForceOrder` component with `before` and `after` members that you can use while sorting instead?
It looks much easier and even more efficient. 🤔

**Reply by hfn92** [reply]
> Elements are stored in a paged vector. move_to isn't that easy to implement actually. 🙂

I see, I just assumed if sorting is already possible such a thing would be easy to do :sweat_smile: 

>However, why not a single ForceOrder component with before and after members that you can use while sorting instead?
It looks much easier and even more efficient. 🤔

I didn't think that would work since `less than` wouldn't be implemented correctly. What should be returned when one of the entities is not one it should be ordered against? I think that would violate the sorting algorithm.

```cpp
esc.sort<ToSort>([&](entity& l, entity& r) 
{
   auto lval = ecs.get<ToSort>(l);
   auto rval = ecs.get<ToSort>(r);
   
   if (auto order = ecs.try_get<ForceOrderBefore>(l); order && order.entity == r) return true;
   if (auto order = ecs.try_get<ForceOrderBefore>(r); order && order.entity == l) return false;

   return lval < rval; 
});
```

Or am I misunderstanding something?

**Reply by hfn92** [reply]
Ok, I guess if an entitiy contains `ForceOrder` I should use the entity referenced in `ForceOrder` to compare?

```cpp
   auto lval = ecs.get<ToSort>(l);
   auto rval = ecs.get<ToSort>(r);
   
   if (auto order = ecs.try_get<ForceOrderBefore>(l)) 
     if (order->entity == r) return true;
     else lval = ecs.get<ToSort>(order->entity)
   if (auto order = ecs.try_get<ForceOrderBefore>(r)) 
     if (order->entity == l) return false;
     else rval = ecs.get<ToSort>(order->entity)

   return lval < rval; 
```

I can try that. But I'm not sure if that really if more performant considering that this is a rare case.

**Reply by skypjack** [reply]
Before we dig a little deeper, let me ask this: what happens if `e1` has a `ForceOrderBefore` on `e2` and `e2` has a `ForceOrderBefore` on `e1`? Because it seems possible to me. In this case, the sorting algorithm would break no matter what. The result is an UB probably, a very subtle one too.

Can I ask you what problem are you trying to solve too? We may be facing an XY-problem and there may be a better solution than what you are attempting.

**Reply by hfn92** [reply]
>Before we dig a little deeper, let me ask this: what happens if e1 has a ForceOrderBefore on e2 and e2 has a ForceOrderBefore on e1? Because it seems possible to me. In this case, the sorting algorithm would break no matter what. The result is an UB probably, a very subtle one too.

Yeah, I didn't think this would work for all cases. That was just more for the general idea.

>Can I ask you what problem are you trying to solve too? We may be facing an XY-problem and there may be a better solution than what you are attempting.

I need to z-sort some entities before rendering. They are being iterated multiple times for different passes when rendering. In some very rare cases I would like to defy the z-order of an entity and render it exactly before another. 

I can think of a couple of other ways to solve this of course, but as they are being iterated by multiple systems, the most convenient way (for me at least :smile:) would be to reorder these entities once after sorting and not have to worry about checking anything after that or pay the cost of `try_get` for this rare <1% case.  Which brings me to my original question :smile:

**Reply by skypjack** [reply]
How does the `ForceOrderBefore` enter the game then? I would expect something like a `Renderable` component with a `z` member instead.

**Reply by hfn92** [reply]
I would add that when creating the entity with its sub entities, each of which have their own components required for rendering.

**Reply by skypjack** [reply]
So, the parent entity has not the `ForceOrderBefore` and no `z` at all while the children have a `ForceOrderBefore` which is a sort of _higher `z`_ if compared with that of the parent?

**Reply by hfn92** [reply]
The parent entity might have it's own graphical components. I'd just want an entity rendered in front of another despite it's z value being lower. The parent child relationship doesn't really matter for the renderer.

**Reply by skypjack** [reply]
I must admit that I don't get exactly all conditions and nuances of your approach.
In any case, from your last sentence you can just have a z and actual z value, the latter is also affected by `ForceOrderBefore` in a separate system. Then you can just sort it rather than trying the path of a post-sort fix of your order.
However, all solutions have the same problem imho. That is, you can have inconsistency in how `ForceOrderBefore` components are assigned to entities and this leads to errors no matter what.
